### PR TITLE
Fix non atomic cache updates

### PR DESCRIPTION
**Pull request by zhangchiqing**
Problem: Adding a block to database requires to insert both headers and payloads to badger db in a transaction. However the updates to the cache are non-atomic, which caused some bugs because of "dirty reads" - reading cached data from an incomplete transaction.

I've included a test case (see comments) to reproduce the problem and confirm this has been fixed.

This PR introduces the minimum changes required to fix it [5506](https://github.com/dapperlabs/flow-go/issues/5506), so that we can focus on the main fix.

If this PR is approved, I will create another PR to clean up other modules.

### CODE REVIEW

File: storage/badger/transaction/tx.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+	err = dbTxn.Commit()
+	if err != nil {
+		return err
+	}
+
+	for _, callback := range tx.callbacks {
+		callback()
+	}
+	return nil
```
Comment (zhangchiqing) [inline storage/badger/transaction/tx.go:38]: This is the main fix. The idea is to create callbacks to update cache after the transaction has been successfully committed.

**Review by jordanschalm — APPROVED** [review]
Nice work 💯 . The approach looks good to me.

**Review by AlexHentschel — APPROVED** [review]
Thanks Leo. This is great.

File: storage/badger/payloads.go [hunk 2]
```diff
@@ -96,6 +97,68 @@ func (p *Payloads) storeTx(blockID flow.Identifier, payload *flow.Payload) func(
...
+	// and execution result is decoupled.
+	resultsById := payload.Results.Lookup()
+	fullReceipts := make([]*flow.ExecutionReceipt, 0, len(payload.Receipts))
+	var err error
+	for _, meta := range payload.Receipts {
+		result, ok := resultsById[meta.ResultID]
+		if !ok {
+			result, err = p.results.ByID(meta.ResultID)
+			if err != nil {
+				if errors.Is(err, badger.ErrKeyNotFound) {
```
[the remark this answers is not in the document]
Comment (zhangchiqing) [reply storage/badger/payloads.go]: Thanks, just fixed, and added the comments [here](https://github.com/onflow/flow-go/pull/697/commits/3383970f5b1058801c7a4402108f2eb544e72841)

**Pull request merged**